scripts/cluster/setup_master_node.go

```diff
@@ -199,18 +215,42 @@ func InstallKnativeServingComponent(stockContainerd string) error {
 		if !utils.CheckErrorWithMsg(err, "Failed to install Knative Serving component!\n") {
 			return err
 		}
-		_, err = utils.ExecShellCmd("kubectl apply -f https://github.com/knative/serving/releases/download/knative-v%s/serving-core.yaml", configs.Knative.KnativeVersion)
-		if !utils.CheckErrorWithTagAndMsg(err, "Failed to install Knative Serving component!\n") {
-			return err
+		
+		if _, err = os.Stat(path.Join(configs.VHive.VHiveRepoPath, path.Join("configs/knative_yamls", "serving-core.yaml"))); err != nil {
+			_, err = utils.ExecShellCmd("kubectl apply -f https://github.com/knative/serving/releases/download/knative-v%s/serving-core.yaml", configs.Knative.KnativeVersion)
+			if !utils.CheckErrorWithTagAndMsg(err, "Failed to install Knative Serving component!\n") {
+				return err
+			}
+		} else {
+			servingCorePath, err := utils.GetVHiveFilePath(path.Join("configs/knative_yamls", "serving-core.yaml"))
+			if err != nil {
+				return err
+			}
+			_, err = utils.ExecShellCmd("cat %s | envsubst | kubectl apply -f -", servingCorePath)
+			if !utils.CheckErrorWithTagAndMsg(err, "Failed to install Knative Serving component!\n") {
+				return err
+			}
 		}
 	} else {
 		_, err := utils.ExecShellCmd("kubectl apply -f %s/serving-crds.yaml", configs.Knative.NotStockOnlyKnativeServingYamlUrlPrefix)
 		if !utils.CheckErrorWithMsg(err, "Failed to install Knative Serving component!\n") {
 			return err
 		}
-		_, err = utils.ExecShellCmd("kubectl apply -f %s/serving-core.yaml", configs.Knative.NotStockOnlyKnativeServingYamlUrlPrefix)
-		if !utils.CheckErrorWithTagAndMsg(err, "Failed to install Knative Serving component!\n") {
-			return err
+
+		if _, err = os.Stat(path.Join(configs.VHive.VHiveRepoPath, path.Join("configs/knative_yamls", "serving-core-firecracker.yaml"))); err != nil {
```

Comment:
This if is redundant, we always should find this file in the folder. I think, you should remove completely the field NotStockOnlyKnativeServingYamlUrlPrefix from configs then.